### PR TITLE
feat: Implement 'open' and 'unconstrained' keywords

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

Resolves #(none)

# Description

## Summary of changes

Adds the 'open' and 'unconstrained' keywords for contract visibility. These keywords are mutually exclusive and functions with neither keyword are considered to be 'secret' by default. It is an error for functions outside of a contract to specify that they are open or unconstrained (though we may allow unconstrained functions in the future).

These keywords currently do nothing until #1033 is merged, when they will only set the is_secret option in the resulting JSON after compiling a contract.

## Dependency additions / changes

None

## Test additions / changes

None

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [x] This PR requires documentation updates when merged.

Once we document contracts, we will need to document that they can be secret (the default, no keyword), `open`, or `unconstrained`.

# Additional context

### CODE REVIEW

File: crates/noirc_frontend/src/ast/expression.rs [hunk 1]
```diff
@@ -318,6 +319,29 @@ pub struct FunctionDefinition {
     pub return_visibility: noirc_abi::AbiVisibility,
 }
 
+/// ContractFunctionType describes the types
```
Comment (vezenovm) [inline crates/noirc_frontend/src/ast/expression.rs:322]: Rename enum in the comment to `ContractVisibility`

**Review by vezenovm — DISMISSED** [review]
Just one nit

**Review by vezenovm — DISMISSED** [review]

File: crates/noirc_frontend/src/hir/resolution/resolver.rs [hunk 2]
```diff
@@ -643,6 +644,17 @@ impl<'a> Resolver<'a> {
         }
     }
 
+    fn handle_contract_visibility(&mut self, func: &NoirFunction) -> ContractVisibility {
+        let mut contract_visibility = func.def.contract_visibility;
+        if !self.in_contract() && contract_visibility != ContractVisibility::Secret {
```
Comment (guipublic) [inline crates/noirc_frontend/src/hir/resolution/resolver.rs:649]: It uses some assumptions about contract visibility, which would need to change if we change the behaviour. I think having an explicit visibility for a function will be more clear, like having an optional contract visibility (which is set only for contracts), or a normal function type.

File: crates/noirc_frontend/src/hir/resolution/resolver.rs [hunk 3]
```diff
@@ -1241,9 +1259,22 @@ mod test {
...
+        let mut def_maps: HashMap<CrateId, CrateDefMap> = HashMap::new();
         let file = FileId::default();
 
+        let mut modules = arena::Arena::new();
+        modules.insert(ModuleData::new(None, ModuleOrigin::File(file), false));
+
+        def_maps.insert(
+            CrateId::dummy_id(),
+            CrateDefMap {
+                root: LocalModuleId(arena::Index::from_raw_parts(0, 0)),
```
Comment (guipublic) [inline crates/noirc_frontend/src/hir/resolution/resolver.rs:1271]: you should use the localmoduleid() function

File: crates/noirc_frontend/src/hir/resolution/resolver.rs [hunk 4]
```diff
@@ -1445,7 +1476,13 @@ mod test {
         }
 
         fn local_module_id(&self) -> LocalModuleId {
-            LocalModuleId::dummy_id()
+            // This is not LocalModuleId::dummy since we need to use this to index into a Vec
+            // later and do not want to push u32::MAX number of elements before we do.
+            LocalModuleId(arena::Index::from_raw_parts(0, 0))
```
Comment (guipublic) [inline crates/noirc_frontend/src/hir/resolution/resolver.rs:1481]: same here; localmoduleid() function

File: crates/noirc_frontend/src/hir/resolution/resolver.rs (hunk 2 above)
Comment (jfecher) [reply crates/noirc_frontend/src/hir/resolution/resolver.rs:649]: I've changed it to `Option<ContractVisibility>`

File: crates/noirc_frontend/src/lexer/token.rs [hunk 5]
```diff
@@ -432,10 +432,12 @@ pub enum Keyword {
     Let,
     Mod,
     Mut,
+    Open,
     Pub,
     String,
     Return,
     Struct,
+    Unconstrained,
```
[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_frontend/src/lexer/token.rs]: I've changed the keyword

**Review by guipublic — APPROVED** [review]

**Pull request merged**